device/cuda/src/clusterization/clusterization_algorithm.cpp

```diff
@@ -0,0 +1,143 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Clusterization include(s)
+#include "traccc/cuda/clusterization/clusterization_algorithm.hpp"
+
+#include "traccc/device/get_prefix_sum.hpp"
+
+// CUDA library include(s).
+#include "cluster_counting.hpp"
+#include "clusters_sum.hpp"
+#include "component_connection.hpp"
+#include "measurements_creation.hpp"
+#include "spacepoint_formation.hpp"
+
+// Vecmem include(s).
+#include <algorithm>
+#include <vecmem/utils/copy.hpp>
+
+namespace traccc::cuda {
+
+clusterization_algorithm::clusterization_algorithm(vecmem::memory_resource &mr)
+    : m_mr(mr) {}
+
+clusterization_algorithm::output_type clusterization_algorithm::operator()(
+    const cell_container_types::host &cells_per_event) const {
+
+    // Number of modules
+    unsigned int num_modules = cells_per_event.size();
+
+    // Vecmem copy object for moving the data between host and device
+    vecmem::copy copy;
+
+    // Get the view of the cells container
+    auto cells_data = get_data(cells_per_event, &m_mr.get());
+    cell_container_types::const_view cells_view(cells_data);
+
+    // Get the sizes of the cells in each module
+    auto cell_sizes = copy.get_sizes(cells_view.items);
+
+    // Get the cell sizes with +1 in each entry for sparse ccl indices buffer
+    // The +1 is needed to store the number of found clusters at the end of
+    // the vector in each module
+    std::vector<std::size_t> cell_sizes_plus(num_modules);
+    std::transform(cell_sizes.begin(), cell_sizes.end(),
+                   cell_sizes_plus.begin(),
+                   [](std::size_t x) { return x + 1; });
```

Comment:
I would personally consider the storage of semantically heterogeneous values in a vector like this to be confusing and bad practice, is there no way to store this in a separate variable?

Reply:
Sorry for the delayed response. I will proceed to refactor the code with PR #205. This will be resolved with it.